# <b>Loan Default Risk Prediction</b>

## <b>Exploratory Data Analysis (EDA)</b>

### <b>Session 1: Dataset Overview</b>

**Objective:**

The objective of this session is to understand the overall structure of the dataset before performing any preprocessing or feature engineering.

In this session, we will:

- Load the dataset using the Data Ingestion Pipeline
- Inspect the dataset structure
- Understand the dimensions of the dataset
- Identify numerical and categorical features
- Check missing values and duplicate records

#### <b>Step 1: Import Required Libraries</b>

We begin by importing the required libraries for data manipulation and loading the dataset.

Notice that we are **not using `pd.read_csv()` directly**. Instead, we load the data through our production Data Ingestion component.

In [1]:
import pandas as pd
import numpy as np

from src.components.data_ingestion import DataIngestion

#### <b>Step 2: Load Dataset</b>

The dataset is loaded using the Data Ingestion pipeline.

Using a centralized ingestion component improves maintainability because any future change in the data source only needs to be updated in one place.

In [2]:
ingestion  = DataIngestion()
df = ingestion.load_data()

[2026-07-22 07:05:28,813] INFO - loan_default_project - Loading dataset from: D:\MachineLearningProjects\Loan_Default_Risk_Prediction\data\raw\application_train.csv
[2026-07-22 07:05:36,723] INFO - loan_default_project - Dataset loaded successfully with shape: (307511, 122)


#### <b>Step 3: Display First Five Records</b>

The `head()` function displays the first five rows of the dataset.

This helps us verify:

- Column names
- Data types
- Sample values
- Overall dataset structure

Instead of printing the complete dataset, which contains more than 300,000 records, we inspect only a small sample.

In [3]:
df.head()

,SK_ID_CURR,TARGET,NAME_CONTRACT_TYPE,CODE_GENDER,FLAG_OWN_CAR,FLAG_OWN_REALTY,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
0,100002,1,Cash loans,M,N,Y,0,202500.0,406597.5,24700.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,1.0
1,100003,0,Cash loans,F,N,N,0,270000.0,1293502.5,35698.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
2,100004,0,Revolving loans,M,Y,Y,0,67500.0,135000.0,6750.0,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0
3,100006,0,Cash loans,F,N,Y,0,135000.0,312682.5,29686.5,...,0,0,0,0,NaN,NaN,NaN,NaN,NaN,NaN
4,100007,0,Cash loans,M,N,Y,0,121500.0,513000.0,21865.5,...,0,0,0,0,0.0,0.0,0.0,0.0,0.0,0.0


##### <b>Observation</b>

- The dataset contains customer demographic and financial information.
- Each row represents one loan application.
- The target variable is **TARGET**.
- The dataset contains both numerical and categorical variables.

#### <b>Step 4: Check Dataset Shape</b>

The shape of the dataset tells us the number of rows and columns available for analysis.

In [4]:
df.shape

(307511, 122)

##### <b>Observation</b>

The dataset contains:

- **307,511 rows**
- **122 columns**

This indicates a relatively large dataset with many features that will require preprocessing before model training.

#### <b>Step 5: Dataset Information</b>

The `info()` function provides:

- Number of rows
- Number of columns
- Non-null values
- Data types
- Memory usage

This is one of the first commands every Data Scientist runs before starting EDA.

In [5]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 307511 entries, 0 to 307510
Columns: 122 entries, SK_ID_CURR to AMT_REQ_CREDIT_BUREAU_YEAR
dtypes: float64(65), int64(41), object(16)
memory usage: 286.2+ MB


##### <b>Observation</b>

From the dataset information we observe:

- Multiple columns contain missing values.
- Both numerical and categorical features are present.
- Most numerical columns are stored as `float64` or `int64`.
- Several categorical columns are stored as `object`.

#### <b>Step 6: Statistical Summary</b>

The `describe()` function summarizes the numerical features.

It provides:

- Count
- Mean
- Standard Deviation
- Minimum
- Quartiles
- Maximum

In [6]:
df.describe()

,SK_ID_CURR,TARGET,CNT_CHILDREN,AMT_INCOME_TOTAL,AMT_CREDIT,AMT_ANNUITY,AMT_GOODS_PRICE,REGION_POPULATION_RELATIVE,DAYS_BIRTH,DAYS_EMPLOYED,...,FLAG_DOCUMENT_18,FLAG_DOCUMENT_19,FLAG_DOCUMENT_20,FLAG_DOCUMENT_21,AMT_REQ_CREDIT_BUREAU_HOUR,AMT_REQ_CREDIT_BUREAU_DAY,AMT_REQ_CREDIT_BUREAU_WEEK,AMT_REQ_CREDIT_BUREAU_MON,AMT_REQ_CREDIT_BUREAU_QRT,AMT_REQ_CREDIT_BUREAU_YEAR
count,307511.000000,307511.000000,307511.000000,3.075110e+05,3.075110e+05,307499.000000,3.072330e+05,307511.000000,307511.000000,307511.000000,...,307511.000000,307511.000000,307511.000000,307511.000000,265992.000000,265992.000000,265992.000000,265992.000000,265992.000000,265992.000000
mean,278180.518577,0.080729,0.417052,1.687979e+05,5.990260e+05,27108.573909,5.383962e+05,0.020868,-16036.995067,63815.045904,...,0.008130,0.000595,0.000507,0.000335,0.006402,0.007000,0.034362,0.267395,0.265474,1.899974
std,102790.175348,0.272419,0.722121,2.371231e+05,4.024908e+05,14493.737315,3.694465e+05,0.013831,4363.988632,141275.766519,...,0.089798,0.024387,0.022518,0.018299,0.083849,0.110757,0.204685,0.916002,0.794056,1.869295
min,100002.000000,0.000000,0.000000,2.565000e+04,4.500000e+04,1615.500000,4.050000e+04,0.000290,-25229.000000,-17912.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
25%,189145.500000,0.000000,0.000000,1.125000e+05,2.700000e+05,16524.000000,2.385000e+05,0.010006,-19682.000000,-2760.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000
50%,278202.000000,0.000000,0.000000,1.471500e+05,5.135310e+05,24903.000000,4.500000e+05,0.018850,-15750.000000,-1213.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,1.000000
75%,367142.500000,0.000000,1.000000,2.025000e+05,8.086500e+05,34596.000000,6.795000e+05,0.028663,-12413.000000,-289.000000,...,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,0.000000,3.000000
max,456255.000000,1.000000,19.000000,1.170000e+08,4.050000e+06,258025.500000,4.050000e+06,0.072508,-7489.000000,365243.000000,...,1.000000,1.000000,1.000000,1.000000,4.000000,9.000000,8.000000,27.000000,261.000000,25.000000


##### <b>Observation</b>

The statistical summary helps us identify:

- Features with large ranges
- Possible outliers
- Skewed distributions
- Missing values (through reduced count)

#### <b>Step 7: Missing Value Analysis</b>

Before cleaning the data, we identify which columns contain missing values.

This helps us decide the appropriate imputation strategy during preprocessing.

In [7]:
df.isnull().sum().sort_values(ascending=False).head(10)

COMMONAREA_AVG              214865
COMMONAREA_MODE             214865
COMMONAREA_MEDI             214865
NONLIVINGAPARTMENTS_MEDI    213514
NONLIVINGAPARTMENTS_MODE    213514
NONLIVINGAPARTMENTS_AVG     213514
FONDKAPREMONT_MODE          210295
LIVINGAPARTMENTS_AVG        210199
LIVINGAPARTMENTS_MEDI       210199
LIVINGAPARTMENTS_MODE       210199
dtype: int64

##### <b>Observation</b>

Several columns contain a high percentage of missing values.

These columns will require careful handling during the feature engineering stage.

At this point, we only identify missing values—we do not remove or impute them.

#### <b>Session 1 Summary</b>

In this session we:

- Successfully loaded the dataset
- Understood the dataset dimensions
- Explored the dataset structure
- Identified numerical and categorical features
- Checked duplicate records
- Identified missing values
- Generated initial business observations

The dataset is now ready for deeper exploratory analysis in the next session.